homeassistant/components/bmw_connected_drive/lock.py

```diff
@@ -27,71 +26,62 @@ async def async_setup_entry(
+
+    entities: list[BMWLock] = []
 
-    if not account.read_only:
-        entities = [
-            BMWLock(account, vehicle, "lock", "BMW lock")
-            for vehicle in account.account.vehicles
-        ]
-        async_add_entities(entities, True)
+    for vehicle in coordinator.account.vehicles:
+        if not coordinator.read_only:
+            entities.append(BMWLock(coordinator, vehicle, "lock", "BMW lock"))
+    async_add_entities(entities)
 
 
 class BMWLock(BMWConnectedDriveBaseEntity, LockEntity):
     """Representation of a BMW vehicle lock."""
 
     def __init__(
         self,
-        account: BMWConnectedDriveAccount,
+        coordinator: BMWDataUpdateCoordinator,
         vehicle: ConnectedDriveVehicle,
         attribute: str,
         sensor_name: str,
     ) -> None:
         """Initialize the lock."""
-        super().__init__(account, vehicle)
+        super().__init__(coordinator, vehicle)
 
         self._attribute = attribute
         self._attr_name = f"{vehicle.name} {attribute}"
         self._attr_unique_id = f"{vehicle.vin}-{attribute}"
         self._sensor_name = sensor_name
         self.door_lock_state_available = DOOR_LOCK_STATE in vehicle.available_attributes
 
+    @property
+    def extra_state_attributes(self) -> dict:
+        """Return entity specific state attributes."""
+        return {"door_lock_state": self.vehicle.status.door_lock_state.value}
+
+    @property
+    def is_locked(self) -> bool | None:
+        """Return true if lock is locked."""
+        _LOGGER.debug(
+            "Updating lock data for '%s' of %s", self._attribute, self.vehicle.name
+        )
+        vehicle_state = self.vehicle.status
+        if not self.door_lock_state_available:
+            return None
+        return vehicle_state.door_lock_state in {
+            LockState.LOCKED,
+            LockState.SECURED,
+        }
+
     def lock(self, **kwargs: Any) -> None:
         """Lock the car."""
-        _LOGGER.debug("%s: locking doors", self._vehicle.name)
-        # Optimistic state set here because it takes some time before the
-        # update callback response
-        self._attr_is_locked = True
+        _LOGGER.debug("%s: locking doors", self.vehicle.name)
+        self.vehicle.remote_services.trigger_remote_door_lock()
         self.schedule_update_ha_state()
-        self._vehicle.remote_services.trigger_remote_door_lock()
 
     def unlock(self, **kwargs: Any) -> None:
         """Unlock the car."""
-        _LOGGER.debug("%s: unlocking doors", self._vehicle.name)
-        # Optimistic state set here because it takes some time before the
-        # update callback response
-        self._attr_is_locked = False
+        _LOGGER.debug("%s: unlocking doors", self.vehicle.name)
+        self.vehicle.remote_services.trigger_remote_door_unlock()
         self.schedule_update_ha_state()
-        self._vehicle.remote_services.trigger_remote_door_unlock()
-
-    def update(self) -> None:
```

Comment:
Override `def _handle_coordinator_update(self) -> None:` instead so you don't have to convert back to using the long hand `@property` method

Reply:
Adjusted.